deoxys/src/modes.rs

```diff
@@ -111,75 +111,71 @@ where
         tweak[8] = nonce[7] << 4;
 
         // Message authentication and encryption
-        if !buffer.is_empty() {
-            tweak[0] = (tweak[0] & 0xf) | TWEAK_M;
 
-            let (data_blocks, tail) = buffer.into_chunks();
-            let data_blocks_len = data_blocks.len();
+        tweak[0] = (tweak[0] & 0xf) | TWEAK_M;
 
-            for (index, data) in data_blocks.into_iter().enumerate() {
-                // Copy block number
-                let tmp = tweak[8] & 0xf0;
-                tweak[8..].copy_from_slice(&(index as u64).to_be_bytes());
-                tweak[8] = (tweak[8] & 0xf) | tmp;
+        let (data_blocks, mut tail) = buffer.into_chunks();
+        let data_blocks_len = data_blocks.len();
 
-                for (c, d) in checksum.iter_mut().zip(data.get_in().iter()) {
-                    *c ^= d;
-                }
+        for (index, data) in data_blocks.into_iter().enumerate() {
+            // Copy block number
+            let tmp = tweak[8] & 0xf0;
+            tweak[8..].copy_from_slice(&(index as u64).to_be_bytes());
+            tweak[8] = (tweak[8] & 0xf) | tmp;
 
-                B::encrypt_inout(data, &tweak, subkeys);
+            for (c, d) in checksum.iter_mut().zip(data.get_in().iter()) {
+                *c ^= d;
             }
 
-            let mut data = tail;
-            let index = data_blocks_len;
-            if !data.is_empty() {
-                // Last block, incomplete
+            B::encrypt_inout(data, &tweak, subkeys);
+        }
 
-                // Copy block number
-                let tmp = tweak[8] & 0xf0;
-                tweak[8..].copy_from_slice(&(index as u64).to_be_bytes());
-                tweak[8] = (tweak[8] & 0xf) | tmp;
+        if tail.is_empty() {
+            // Tag computing without last block
+            tweak[0] = (tweak[0] & 0xf) | TWEAK_TAG;
 
-                // Last block checksum
-                tweak[0] = (tweak[0] & 0xf) | TWEAK_M_LAST;
+            let tmp = tweak[8] & 0xf0;
+            tweak[8..].copy_from_slice(&((buffer_len / 16) as u64).to_be_bytes());
+            tweak[8] = (tweak[8] & 0xf) | tmp;
 
-                let mut block = Block::default();
-                block[0..data.len()].copy_from_slice(data.get_in());
+            B::encrypt_inout((&mut checksum).into(), tweak.as_ref(), subkeys);
 
-                block[data.len()] = 0x80;
+            for (t, c) in tag.iter_mut().zip(checksum.iter()) {
+                *t ^= c;
+            }
```

Comment:
@tarcieri 
I tried to replace this part with `tag ^= checksum;`, but I see that `Array` does not implement the bit op traits. Was it intentional to mirror the built-in arrays?

Reply:
We could support that, but yeah, core arrays don't, so if we ever try to migrate that back it's one more thing to remove.

FWIW I asked about adding arithmetic op impls to core arrays on IRLO and people did not seem very receptive.

Reply:
Yeah, I've seen your post and I think I even asked for the same thing somewhere in the rust-lang/rust repo a long time ago with the similar result...